### ISSUE TITLE
Logitech Wrapper patch depends on LGS to be installed

### ISSUE THREAD

**Issue by incryptx** [antonpup/Aurora#2101]
I spent hours trying to get Aurora 0.7.9/0.8.1 to run correctly, to get the Logitech Device Manager to connect and installing the Aurora Wrapper Patch. Of course, without LGS installed, the patch fails because the path and .dll do not exist. As well, the integration of Aurora with GHUB also fails to add Aurora to the integration list of apps.

Now, I installed LGS, and magically, all this stuff started working again. The Logitech Device Manager connects, the Aurora Wrapper Patch works and the integration of Aurora with GHUB works.

So, it appears that the author has adapted Aurora to run with GHUB, but because GHUB is missing it's own SDK, it does not work that well. I suggest rewriting the code for updating the LGS SDK LED Patch (Aurora Wrapper Patch) to do it's magic on the GHUB "sdk_legacy_sdk_led_x64.dll" and it's x86 version instead of trying to patch a non-existent .dll which does not exist unless you have LGS installed. Why the Logitech Device Manager connects with LGS installed, I have no clue. Why it is intermittent at best with GHUB is unknown.

I think the Aurora application needs some serious code updating to work with GHUB and leave it's dependence on LGS behind. I thank the author for providing this excellent software for LGS when this was the only thing available from Logitech, but now that GHUB supersedes it, it's time to update Aurora to work well with GHUB.

I think most of the existing bugs are due to Aurora's continued dependence on LGS.

Thanks

**Comment by Bluscream** [comment]
I would even prefer if aurora would have its own driver for logitech hardware so we can completely ignore that GHIB bs alltogether. It just adds unnessecary overhead

**Comment by diogotr7** [comment]
This is incorrect. The wrapper patch is not needed **at all** for Aurora to communicate with logitech devices. All the patch does is redirect lighting from games that target Logitech's LIGHTSYNC SDK to aurora so that it can be used with other devices. For example, you can get Battlefield 1's lighting on razer and corsair devices. 

When it comes to devices, Aurora can't use the traditional, Logitech-suggested method of using their `LogitechLedEnginesWrapper.dll` (which, as far as I can tell, only redirects the calls made to it to either the GHUB dll or the LGS one). This is because Aurora 'exploits' this dependency, bypassing the need to call this dll, opting instead to call the GHUB / LGS dll's directly. 

This has the advantage of being able to redirect calls to `LogitechLedEnginesWrapper.dll` to Aurora's custom wrapper dll instead.

I do think the current approach of this wrapper is not ideal for two reasons:
1. It's not obvious to the end user what the wrapper *is* and what it does. This should be clarified in the UI. Should also maybe be hidden behind an 'Advanced' menu because of point #2.
2. This patch makes all apps that interact with Logitech's SDK stop working if aurora isn't running / installed. This happens because the dll aurora uses does **not** handle sending the lighting to LGS / GHUB if it doesn;t detect aurora.

I see two main  ways of solving this:
Aurora can keep its own dll, with a more descriptive name such as `AuroraLogitechWrapper.dll` or something similar, in its own folder. Rewriting the official dll's (and thus depending on LGS being installed...) is stupid.
Then:
1. Aurora can either have buttons to patch / unpatch the registry path that tells `LogitechLedEnginesWrapper.dll`  what dll to use (could have 3 buttons: 'Restore GHUB', 'Restore LGS', 'Apply Patch')
2. Or Aurora can patch its own directory when it opens, restoring it to what it was when it closes. This has the advantage of keeping both the wrapper behaviour and restoring the original functionality when aurora closes. It also has the disadvantage of leaving loose ends if aurora closes unexpectedly and is prone to more error.

I would like feedback on this.

**Comment by incryptx** [comment]
I'm very confused by your statements here, but I will defer to your estimation as you are the developer.

You state in the first sentence, that "The wrapper patch is not needed at all for Aurora to communicate with logitech devices. All the patch does is redirect lighting from games that target Logitech's LIGHTSYNC SDK to aurora so that it can be used with other devices."

Then you go on to state "the wrapper this, and the wrapper that". So either your speaking about two different versions of this "wrapper" or the one used by LGS IS NEEDED.

As I stated, it appears Aurora was written specifically for LGS because it contained an SDK for which Logitech encouraged developers like yourself to build applications, which you did. (Thank You, it is an excellent piece of software).

Then Logitech changed to GHUB, which did not contain an SDK, and converted the .dlls used by LGS to "legacy SDK .dlls" and the GHUB directory contains these legacy .dlls for ARX, Steeringwheel, LED, ect., for which Aurora does not seem to use or depend on.

All I can say is that I installed Aurora without LGS,, and it's device manager does not connect for me. Nor does the wrapper patch work (needed or not, obsolete) which confuses folks.

In conclusion, I have Aurora working now with LGS installed, even though I don't use it or allow it to run at startup. 

All I am saying is that Aurora was originally built for LGS and worked great. Then Logitech changed to GHUB, and Aurora was (for the lack of a better term) hacked to work with GHUB, but without it's own SDK, there are limitations to what you can do with GHUB, given Logitech's reluctance to give developers access to the low level drivers and how to interface with them now.

So, Aurora is a nice piece of software and I enjoy it, but due to it's shortcomings with GHUB, it's hard to get it to work without LGS and some dependency.

Not the feedback you wanted, but as a developer myself, having worked at M$ for ten years, I often had to make third party software purchased by M$ and not built to use current drivers and software, work for the masses. It was often a hack. Until M$ allowed the rebuilding of the software from the ground up, users often were left with constantly patched software with workarounds.

**Comment by diogotr7** [comment]
I'll try and clarify what Aurora does to support both LGS and GHUB.

As i said above, Aurora ships with 2 different dll's, one for each program:
https://github.com/antonpup/Aurora/tree/master/Project-Aurora/Project-Aurora/Logi

The Dll contained within the GHUB folder in the above directory is simply the `sdk_legacy_sdk_led_x64.dll` you mentioned above, but renamed.

Aurora tries to 'guess' which of the two programs you're using by scanning for the existance of the GHUB install folder. 
If it finds, it, it will use the GHUB dll. If it doesn't, it uses the LGS one. After this check, everything should work correctly. I've used Aurora exclusively with G HUB installed without any issues (i did install LGS again because it tends to be more stable in my setup, but that's beside the point).

You can also override this behaviour in Aurora's Device Manager (click the 'View Options' button in the Logitch row).

The most common issue users usually have with GHUB support is GHUB's slow startup time. By default, on my system and various others, if both Aurora and GHUB are set to open on startup, Aurora opens before GHUB has a chance to completely initialize. This can be 'fixed' (read: worked around...) by adding a delay to Aurora's startup, so that it only tries to initialize once GHUB is ready for it. This can be done manually in the Windows Task Scheduler or through the General Settings in Aurora.


I decided to expand on the 'wrapper' portion of the issue more than i did on the Logitech device connection one above because it seemed relevant, sorry about that. I hope my answer clarifies things.

**Comment by incryptx** [comment]
Three notes.

1) I uninstall LGS and the Aurora Device Manager Logitech Driver will not stay connected no matter what I do. Without LGS, the Logitech Driver starts briefly then stops, but will not stay connected. With LGS installed, The Logitech Driver stays connected without fail.

2) Every time GHUB updates, I loose the integration with Aurora.exe with no way to get it back or install it again. I don't know how Aurora.exe gets installed into GHUB's integration menu, but once it's gone, the only way to get it back is to install LGS and Aurora from scratch. Once Aurora. exe is integrated into GHUB, things work. 

3) I uninstalled GHUB and used Aurora with LGS only as a test. Things worked great. I think this proves a dependency on LGS and that it was written for LGS. I would say that Aurora is not compatable with GHUB, and trying to make it work with GHUB is more hassle than it's worth.

GHUB seems to cause problems for Aurora. Aurora was written for LGS and uses it's SDK driver.

What is needed is a suitable driver written specifically for GHUB, but without an SDK for GHUB and Logitech opting not to provide developers with any interface to GHUB's drivers, your left with trying to come up with a work-around.

**Comment by incryptx** [comment]
That clarifies things abit. Thanks. There must be a way to move the LGS DLL to GHUB and use it exclusively?

Yes, I have run into the conflict between Aurora and GHUB startup and have added a delay in Aurora to prevent that.

I fail to see why one would want to override the DLL when both are there, you can either override the LGS DLL or the GHUB DLL, but why?

I downloaded the source code for the project and was looking at the code. In my case, for whatever reason, Aurora finds the LGS.DLL before it finds the GHUB.DLL and with LGS uninstalled, it seems not to use the GHUB.DLL and this causes it not to connect. I was going to rename and move some DLL's around, but decided to look at the GHUB Logitech SDK instead.

It looks like Logitech offered an SDK for LGS and had everything in one SDK package. But for GHUB,, they have broken everything into five SDK packages. The one that interest me was the LED Illumination SDK which includes some new DLL's.

https://www.logitechg.com/en-gb/innovation/developer-lab.html

I have not looked at them yet, as my machine does not have a build environment or IDE on it and I have to download and configure my machine to build in Windows. Right now I have a build environment on Linux and Python for anther project, but as soon as I get some time, I will examine the GHUB SDK's DLL package and see what I can do to make Aurora more compatible and friendly with GHUB.

Again, you seem to contradict yourself when you say: _"I've used Aurora exclusively with G HUB installed without any issues (i did install LGS again because it tends to be more stable in my setup, but that's beside the point._

You state Aurora will run with no problems with GHUB, yet you have LGS installed because it's more stable. I think that is exactly my point. LGS is more stable than using GHUB alone, and therefore, in lies Aurora's dependence for stability.
